### ISSUE TITLE
Cannot change the setting on a read-only connection while using Select

### ISSUE THREAD

**Issue by Onajk** [ClickHouse/clickhouse-java#1445]
Hi,
I'm facing a problem of running Select query after upgrading CH driver.
CH version: `22.8.15.25`
We were using: `"ru.yandex.clickhouse" % "clickhouse-jdbc" % "0.3.0"`
I have switched to: `"com.clickhouse" % "clickhouse-jdbc" % "0.4.6"`

With 0.3.0 driver, everything was working fine. After upgrading I'm getting exception:
`java.sql.SQLException: Cannot change the setting on a read-only connection`

In logs I can see that before exception there was a query:
`[info] 2023-09-07 17:59:23,669 DEBUG [pool-1-thread-1] [com.clickhouse.client.http.ClickHouseHttpClient] Query: select currentUser() user, timezone() timezone, version() version, toUInt8(ifnull((select value from system.settings where name = 'readonly'), '0')) as readonly, toInt8(ifnull((select value from system.settings where name = 'throw_on_unsupported_query_inside_transaction'), '-1')) as throw_on_unsupported_query_inside_transaction, (ifnull((select value from system.settings where name = 'wait_changes_become_visible_after_commit_mode'), '')) as wait_changes_become_visible_after_commit_mode,toInt8(ifnull((select value from system.settings where name = 'implicit_transaction'), '-1')) as implicit_transaction, toUInt64(ifnull((select value from system.settings where name = 'max_insert_block_size'), '0')) as max_insert_block_size, toInt8(ifnull((select value from system.settings where name = 'allow_experimental_lightweight_delete'), '-1')) as allow_experimental_lightweight_delete, (ifnull((select value from system.settings where name = 'custom_jdbc_config'), '')) as custom_jdbc_config FORMAT RowBinaryWithNamesAndTypes`

I have a read-only account for ClickHouse which I'm reading from but as far as I know, Select needs only read-only permission. Could someone help me understand what is wrong with new driver?

**Comment by zhicwu** [comment]
Hi @Onajk, what's exact query was working fine? And the value of `readonly` setting? The implementation should follow https://clickhouse.com/docs/en/operations/settings/permissions-for-queries.

**Comment by Onajk** [comment]
I have a lot of different queries but looks like the problem isn't related strictly to those queries. I have checked logs of ClickHouse server and I can see this:

> 07:14:51  Executing query:
          -- 8< --
          select name, auth_type,
                   host_ip, host_names, host_names_regexp, host_names_like,
                   default_roles_all, default_roles_list, default_roles_except
              from system.users
          -- >8 --
07:14:52  Code: 497. DB::Exception: readonly: Not enough privileges. To execute this query it's necessary to have grant SHOW USERS ON *.*. (ACCESS_DENIED) (version 22.8.15.25.altinitystable (altinity build))
          , server ClickHouseNode [uri=https://cit1-ach.evolution-test.altinity.cloud:8443/cit1_analytical, options={sslmode=none,custom_http_params=session_id=DataGrip_aae15476-5c06-420f-80e5-995ba2706128}]@-127048698
07:14:52  execution time: 107 ms
07:14:52  Retrieving server objects
07:14:52  Executing query:
          -- 8< --
          select name from system.roles
          -- >8 --
07:14:52  Code: 497. DB::Exception: readonly: Not enough privileges. To execute this query it's necessary to have grant SHOW ROLES ON *.*. (ACCESS_DENIED) (version 22.8.15.25.altinitystable (altinity build))
          , server ClickHouseNode [uri=https://cit1-ach.evolution-test.altinity.cloud:8443/cit1_analytical, options={sslmode=none,custom_http_params=session_id=DataGrip_aae15476-5c06-420f-80e5-995ba2706128}]@-127048698
07:14:52  execution time: 104 ms

I have no idea where those queries come from but looks like they are called before each of my query.

**Comment by Onajk** [comment]
When I run `SHOW GRANTS;`
Response is:

> GRANT SHOW, SELECT, INSERT, ALTER, CREATE, DROP, TRUNCATE, OPTIMIZE, BACKUP, KILL QUERY, KILL TRANSACTION, MOVE PARTITION BETWEEN SHARDS, SYSTEM, dictGet, INTROSPECTION, SOURCES, CLUSTER ON *.* TO readonly

`readonly` is my user

**Comment by Onajk** [comment]
Running both: `SHOW USERS;` and `SHOW ROLES;` on CH gives me `Not enough privileges.` exception. Was driver changed in the way that now I need also those?

**Comment by zhicwu** [comment]
> select currentUser() user, timezone() timezone, version() version, toUInt8(ifnull((select value from system.settings where name = 'readonly'), '0')) as readonly, toInt8(ifnull((select value from system.settings where name = 'throw_on_unsupported_query_inside_transaction'), '-1')) as throw_on_unsupported_query_inside_transaction, (ifnull((select value from system.settings where name = 'wait_changes_become_visible_after_commit_mode'), '')) as wait_changes_become_visible_after_commit_mode,toInt8(ifnull((select value from system.settings where name = 'implicit_transaction'), '-1')) as implicit_transaction, toUInt64(ifnull((select value from system.settings where name = 'max_insert_block_size'), '0')) as max_insert_block_size, toInt8(ifnull((select value from system.settings where name = 'allow_experimental_lightweight_delete'), '-1')) as allow_experimental_lightweight_delete, (ifnull((select value from system.settings where name = 'custom_jdbc_config'), '')) as custom_jdbc_config FORMAT RowBinaryWithNamesAndTypes

This query is from the JDBC driver when initializing an connection. If you want to turn it off, you may specify two options `server_time_zone` and `server_version` in either connection string or properties.

> select name, auth_type,
> host_ip, host_names, host_names_regexp, host_names_like,
> default_roles_all, default_roles_list, default_roles_except
> from system.users

This looks like generated by your database client, DataGrip. My license expired so I cannot test, perhaps it treat old and new drivers differently.

> Running both: SHOW USERS; and SHOW ROLES; on CH gives me Not enough privileges. exception. Was driver changed in the way that now I need also those?

Did you upgrade server or change user privilege recently? This has nothing to do with the driver, you can reproduce the issue using the native command line. If you want the readonly user can access system tables, you have to grant access specifically and for security reason, apply row level security so it won't see everything in the table.

**Comment by Onajk** [comment]
> This query is from the JDBC driver when initializing an connection. If you want to turn it off, you may specify two options server_time_zone and server_version in either connection string or properties.

After upgrading driver, CH started using local time zone of my computer as server so to have the same results as on actual server I have added:
`use_server_time_zone = false,  use_time_zone = "UTC"`
I have now changed it to `server_time_zone = "UTC"` as it sounds even better.
About `server_version`. There are actually two versions which we are using currently.
`21.8.12.29.altinitystable` - which is used on actual server
`22.8.15.25.altinitystable` - which is used on local tests and will be used on server in the near future
Now, the unexpected part.
I have added property `server_version = "21.8.12.29.altinitystable"` and it's working as before but the same situation happened with `server_version = "22.8.15.25.altinitystable"` so looks like it's not related to actual version but more like with new driver I have to specify it in properties in order to require the same permissions for my queries.
